Content.Shared/Turrets/SharedDeployableTurretSystem.cs

```diff
@@ -0,0 +1,170 @@
+    {
+        if (enabled && ent.Comp.Broken)
+        {
+            if (user != null)
+                _popup.PopupClient(Loc.GetString("deployable-turret-component-is-broken"), ent, user.Value);
+
+            return false;
+        }
+
+        if (enabled && !HasAmmo(ent))
+        {
+            if (user != null)
+                _popup.PopupClient(Loc.GetString("deployable-turret-component-no-ammo"), ent, user.Value);
+
+            return false;
+        }
+
+        SetState(ent, enabled, user);
+
+        return true;
+    }
+
+    protected virtual void SetState(Entity<DeployableTurretComponent> ent, bool enabled, EntityUid? user = null)
+    {
+        if (ent.Comp.Enabled == enabled)
+            return;
+
+        // Close the wires panel UI on activation
+        if (enabled && TryComp<WiresPanelComponent>(ent, out var wires))
+            _wires.TogglePanel(ent, wires, false);
+
+        // Update status
+        ent.Comp.Enabled = enabled;
+
+        // Determine how much time is remaining in the current animation and the one next in queue
+        var animTimeRemaining = MathF.Max((float)(ent.Comp.AnimationCompletionTime - _timing.CurTime).TotalSeconds, 0f);
+        var animTimeNext = ent.Comp.Enabled ? ent.Comp.DeploymentLength : ent.Comp.RetractionLength;
+
+        ent.Comp.AnimationCompletionTime = _timing.CurTime + TimeSpan.FromSeconds(animTimeNext + animTimeRemaining);
+
+        // Change the turret's damage modifiers
+        if (TryComp<DamageableComponent>(ent, out var damageable))
+        {
+            var damageSetID = ent.Comp.Enabled ? ent.Comp.DeployedDamageModifierSetId : ent.Comp.RetractedDamageModifierSetId;
+            _damageable.SetDamageModifierSetId(ent, damageSetID, damageable);
+        }
+
+        // Change the turret's fixtures
+        if (ent.Comp.DeployedFixture != null &&
+            TryComp(ent, out FixturesComponent? fixtures) &&
+            fixtures.Fixtures.TryGetValue(ent.Comp.DeployedFixture, out var fixture))
+        {
+            _physics.SetHard(ent, fixture, ent.Comp.Enabled);
+        }
+
+        // Display message
+        var msg = ent.Comp.Enabled ? "deployable-turret-component-activating" : "deployable-turret-component-deactivating";
+        _popup.PopupClient(Loc.GetString(msg), ent, user);
+    }
+
+    public bool HasAmmo(Entity<DeployableTurretComponent> ent)
+    {
+        if (TryComp<ProjectileBatteryAmmoProviderComponent>(ent, out var projectilebatteryAmmo) &&
+            (projectilebatteryAmmo.Shots > 0 || ent.Comp.Powered))
+            return true;
+
+        if (TryComp<HitscanBatteryAmmoProviderComponent>(ent, out var hitscanBatteryAmmo) &&
+            (hitscanBatteryAmmo.Shots > 0 || ent.Comp.Powered))
+            return true;
+
+        if (TryComp<BallisticAmmoProviderComponent>(ent, out var ballisticAmmo) &&
+            ballisticAmmo.Count > 0)
+            return true;
+
+        return false;
+    }
```

Comment:
This is really hacky and not ECS as it's just going to break if any new components get used, it should be using events to check.